File: app/src/main/java/com/example/hellofriend/Activities/ChatActivity.java
package com.example.hellofriend.Activities;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.hellofriend.Adapters.ChatAdapter;
import com.example.hellofriend.Models.Message1;
import com.example.hellofriend.R;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executors;

public class ChatActivity extends AppCompatActivity {

    private RecyclerView chatRecyclerView;
    private ChatAdapter chatAdapter;
    private ArrayList<Message1> messageList; // Updated to Message1
    private EditText editText;
    private ImageButton sendButton;

    private FirebaseFirestore db;
    private FirebaseAuth mAuth;

    private String userId; // The recipient's userId
    private String userName; // The recipient's userName
    private String currentUserId; // The sender's userId

    private MessageDao messageDao;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);

        // Retrieve user details from intent
        userId = getIntent().getStringExtra("userId");
        userName = getIntent().getStringExtra("userName");

        chatRecyclerView = findViewById(R.id.chat_recycler);
        editText = findViewById(R.id.editT);
        sendButton = findViewById(R.id.send_message);

        mAuth = FirebaseAuth.getInstance();
        db = FirebaseFirestore.getInstance();

        currentUserId = mAuth.getCurrentUser().getUid(); // Get the sender's userId
        messageList = new ArrayList<>();
        chatAdapter = new ChatAdapter(messageList); // Use Message1 here

        chatRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        chatRecyclerView.setAdapter(chatAdapter);

        setTitle("Chat with " + userName);

        // Initialize Room Database
        messageDao = AppDatabase.getInstance(this).messageDao();

        // Load Messages
        loadMessages();

        sendButton.setOnClickListener(v -> sendMessage());
    }

    private void loadMessages() {
        // Load messages from Room Database
        LiveData<List<Message1>> liveMessages = messageDao.getChatMessages(currentUserId, userId);
        liveMessages.observe(this, messages -> {
            messageList.clear();
            messageList.addAll(messages);
            chatAdapter.notifyDataSetChanged();
        });

        // Sync messages with Firestore if network is available
        if (isNetworkAvailable()) {
            syncMessagesWithFirestore();
        }
    }

    private void syncMessagesWithFirestore() {
        db.collection("Messages")
                .whereEqualTo("userId", currentUserId)
                .whereEqualTo("recipientId", userId)
                .get()
                .addOnSuccessListener(querySnapshot -> Executors.newSingleThreadExecutor().execute(() -> {
                    for (var doc : querySnapshot.getDocuments()) {
                        Message1 message = doc.toObject(Message1.class);

                        if (message != null) {
                            // Insert message into Room database if it does not already exist
                            messageDao.insertMessage(message);
                        }
                    }
                }))
                .addOnFailureListener(e -> {
                    Toast.makeText(this, "Failed to sync messages", Toast.LENGTH_SHORT).show();
                });
    }

    private void sendMessage() {
        String messageText = editText.getText().toString();
        if (!messageText.isEmpty()) {
            String currentUserName = mAuth.getCurrentUser().getDisplayName();

            // Create a new Message1 object to be inserted into Room Database
            Message1 message = new Message1();
            message.setText(messageText);
            message.setUserId(currentUserId);
            message.setRecipientId(userId);
            message.setUserName(currentUserName);
            message.setTimestamp(System.currentTimeMillis());

            // Insert into Room Database on a background thread
            Executors.newSingleThreadExecutor().execute(() -> messageDao.insertMessage(message));

            // Sync with Firestore if network is available
            if (isNetworkAvailable()) {
                Map<String, Object> messageMap = new HashMap<>();
                messageMap.put("text", messageText);
                messageMap.put("userId", currentUserId);
                messageMap.put("recipientId", userId);
                messageMap.put("userName", currentUserName);
                messageMap.put("timestamp", System.currentTimeMillis());

                db.collection("Messages").add(messageMap);
            }

            editText.setText(""); // Clear the input field
        } else {
            Toast.makeText(this, "Message cannot be empty", Toast.LENGTH_SHORT).show();
        }
    }

    private boolean isNetworkAvailable() {
        ConnectivityManager connectivityManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
    }
}

File: app/src/main/java/com/example/hellofriend/Activities/ChatActivity.kt
package com.example.hellofriend.Activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.EditText
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.hellofriend.Adapters.ChatAdapter
import com.example.hellofriend.MainActivity
import com.example.hellofriend.Models.Message1
import com.example.hellofriend.R
import com.example.hellofriend.ViewModel.ChatViewModel
import com.google.firebase.auth.FirebaseAuth
import com.squareup.picasso.Picasso

class ChatActivity : AppCompatActivity() {
    private lateinit var chatRecyclerView: RecyclerView
    private lateinit var chatAdapter: ChatAdapter
    private lateinit var messageList: ArrayList<Message1>
    private lateinit var sendButton: ImageButton
    private lateinit var backBtn: ImageView
    private lateinit var recipientName: TextView
    private lateinit var imageUser: ImageView
    private lateinit var editText: EditText

    private lateinit var mAuth: FirebaseAuth
    private lateinit var currentUserId: String
    private lateinit var userId: String
    private lateinit var userName: String
    private var recipientImageUrl: String? = null

    private lateinit var chatViewModel: ChatViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)

        // Initialize Firebase Auth
        mAuth = FirebaseAuth.getInstance()
        if (mAuth.currentUser == null) {
            Toast.makeText(this, "User is not logged in", Toast.LENGTH_SHORT).show()
            finish()
            return
        }
        currentUserId = mAuth.currentUser!!.uid

        // Get intent data
        userId = intent.getStringExtra("userId") ?: ""
        userName = intent.getStringExtra("name") ?: ""
        recipientImageUrl = intent.getStringExtra("recipientImageUrl")

        if (userId.isEmpty() || userName.isEmpty()) {
            Toast.makeText(this, "Chat recipient information is missing", Toast.LENGTH_SHORT).show()
            finish()
            return
        }
        initializeViews()
        setUpRecipientDetails()
        setUpRecyclerView()
        initializeViewModel()
        setUpListeners()
    }

    private fun initializeViews() {
        chatRecyclerView = findViewById(R.id.chat_recycler)
        editText = findViewById(R.id.editT)
        sendButton = findViewById(R.id.send_message)
        backBtn = findViewById(R.id.icon_right)
        recipientName = findViewById(R.id.name_chat)
        imageUser = findViewById(R.id.image_user)
    }

    private fun setUpRecipientDetails() {
        recipientName.text = userName
        if (!recipientImageUrl.isNullOrEmpty()) {
            Picasso.get().load(recipientImageUrl).placeholder(R.drawable.ic_me).into(imageUser)
        } else {
            imageUser.setImageResource(R.drawable.ic_me)
        }
    }

    private fun setUpRecyclerView() {
        messageList = ArrayList()
        chatAdapter = ChatAdapter(messageList)
        chatRecyclerView.apply {
            layoutManager = LinearLayoutManager(this@ChatActivity)
            adapter = chatAdapter
        }
    }

    private fun initializeViewModel() {
        chatViewModel = ViewModelProvider(this)[ChatViewModel::class.java]

        // Observe messages in real-time
        chatViewModel.getMessagesLiveData(currentUserId, userId).observe(this, Observer { messages ->
            if (messages != null&& messages.isNotEmpty()) {
                messageList.clear()
                messageList.addAll(messages)
                chatAdapter.notifyDataSetChanged()
                chatRecyclerView.post {
                    chatRecyclerView.smoothScrollToPosition(messageList.size - 1)
                }
            }
        })
    }

    private fun setUpListeners() {
        sendButton.setOnClickListener { sendMessage() }
        backBtn.setOnClickListener { navigateBack() }
    }

    private fun sendMessage() {
        val messageText = editText.text.toString().trim()
        if (messageText.isEmpty()) {
            Toast.makeText(this, "Message cannot be empty", Toast.LENGTH_SHORT).show()
            return
        }

        val message = Message1(
            text = messageText,
            userId = currentUserId,
            recipientId = userId,
            userName = userName,
            firestoreTimestamp = com.google.firebase.Timestamp.now()
        )

        // Send the message using ViewModel
        chatViewModel.sendMessage(message)

        // Clear the EditText
        editText.text.clear()
    }

    private fun navigateBack() {
        val lastMessage = messageList.lastOrNull()
        val lastMessageText = lastMessage?.text ?: "No messages yet"
        val lastMessageTimestamp = lastMessage?.firestoreTimestamp?.toDate()?.time ?: 0

        Log.d("ChatActivity", "Navigating back with Last Message: $lastMessageText, Timestamp: $lastMessageTimestamp")

        val resultIntent = Intent(this, MainActivity::class.java).apply {
            putExtra("lastMessage", lastMessageText)
            putExtra("lastMessageTimestamp", lastMessageTimestamp)
            putExtra("userId", intent.getStringExtra("userId"))
        }
        startActivity(resultIntent)
        finish()
    }



}
